File: RxTango/python/examples/poll_attribute.py
"""Poll a Tango attribute at a fixed interval — no loop, no thread.

Mirrors the Java pattern:
    Flowable.interval(ms).flatMapSingle(read_attribute)

Usage:
    python poll_attribute.py [device] [attribute] [interval-ms]

    device       defaults to tango://localhost:10000/sys/tg_test/1
    attribute    defaults to double_scalar
    interval-ms  defaults to 500
"""

import asyncio
import sys
from datetime import timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import read_attribute


async def main() -> None:
    device      = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"
    attr        = sys.argv[2] if len(sys.argv) > 2 else "double_scalar"
    interval_ms = int(sys.argv[3]) if len(sys.argv) > 3 else 500

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)

    print(f"Polling {device}/{attr} every {interval_ms} ms  (Ctrl+C to stop)\n")
    print(f"  {'value':>20}")
    print("  " + "-" * 22)

    rx.interval(timedelta(milliseconds=interval_ms), scheduler=scheduler).pipe(
        ops.flat_map(lambda _: read_attribute(device, attr)),
    ).subscribe(
        on_next=lambda v: print(f"  {v:>+20.6f}"),
        on_error=lambda e: print(f"  ERROR: {e}", file=sys.stderr),
        scheduler=scheduler,
    )

    await asyncio.Future()  # run until Ctrl+C


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n  stopped.")

File: RxTango/cpp/examples/poll_attribute.cpp
/**
 * Poll a Tango attribute at a fixed interval — no loop, no thread.
 *
 * Mirrors the Java pattern:
 *   Flowable.interval(ms).flatMapSingle(read_attribute)
 *
 * Mirrors Python's poll_attribute.py and Java's PollAttribute.java.
 *
 * Usage:
 *   ./poll_attribute [device] [attribute] [interval-ms]
 *   defaults: sys/tg_test/1  double_scalar  500
 */

#include <chrono>
#include <csignal>
#include <iostream>
#include <string>

#include <rxcpp/rx.hpp>
#include <rxtango/rxtango.hpp>

static volatile sig_atomic_t g_running = 1;

int main(int argc, char* argv[]) {
    const std::string device      = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";
    const std::string attr        = argc > 2 ? argv[2] : "double_scalar";
    const int         interval_ms = argc > 3 ? std::stoi(argv[3]) : 500;

    std::signal(SIGINT, [](int) { g_running = 0; });

    std::cout << "Polling " << device << "/" << attr
              << " every " << interval_ms << " ms  (Ctrl+C to stop)\n\n"
              << "  " << std::setw(20) << "value" << "\n"
              << "  " << std::string(22, '-') << "\n";

    // Interval on a new-thread scheduler; flat_map into single-shot reads.
    // Mirrors: Flowable.interval(ms).flatMapSingle(i -> read)
    auto sub = rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
        .flat_map([device, attr](long) {
            return rxtango::read_attribute<double>(device, attr);
        })
        .subscribe(
            [](double v) { std::cout << "  " << std::setw(+20) << v << "\n"; },
            [](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
            }
        );

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    sub.unsubscribe();
    std::cout << "\n  stopped.\n";
    return 0;
}
